app/src/main/java/io/pumpkinz/pumpkinreader/service/DataSource.java

```diff
@@ -0,0 +1,176 @@
+package io.pumpkinz.pumpkinreader.service;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+
+import java.util.ArrayList;
+import java.util.Dictionary;
+import java.util.Hashtable;
+import java.util.List;
+
+import io.pumpkinz.pumpkinreader.etc.Constants;
+import io.pumpkinz.pumpkinreader.model.News;
+import io.pumpkinz.pumpkinreader.util.Util;
+import rx.Observable;
+import rx.functions.Action1;
+import rx.functions.Func1;
+
+
+public class DataSource {
+
+    private Context ctx;
+
+    public DataSource(Context ctx) {
+        this.ctx = ctx;
+    }
+
+    /**
+     * Get Top News from. Maximum because the JSON retrieved
+     * might not be able to be instantiated into News object due to incomplete JSON.
+     *
+     * @param from      first index to be returned
+     * @param count     maximum number of News to be retrieved
+     * @param isRefresh whether we should request Top News from API or look up the SharedPreferences first
+     * @return an Observable of List of News
+     */
+    public Observable<List<News>> getTop(final int from, final int count, boolean isRefresh) {
+        return listTop(isRefresh)
+                .compose(new NewsTransformer(from, count));
+    }
+
+    public Observable<List<News>> getNew(final int from, final int count, boolean isRefresh) {
+        return listNew(isRefresh)
+                .compose(new NewsTransformer(from, count));
+    }
+
+    private Observable<List<Integer>> listTop(boolean isRefresh) {
+        List<Integer> retval = getNewsFromSp(Constants.TOP_FILE_SP, Constants.TOP_VAL_SP, ctx);
+
+        if (retval.isEmpty() || isRefresh) {
+            return RestClient.service().listTop()
+                    .doOnNext(new putToSpAction(ctx, Constants.TOP_FILE_SP, Constants.TOP_VAL_SP));
+        } else {
+            return Observable.just(retval);
+        }
+    }
+
+    private Observable<List<Integer>> listNew(boolean isRefresh) {
+        List<Integer> retval = getNewsFromSp(Constants.NEW_FILE_SP, Constants.NEW_VAL_SP, ctx);
+
+        if (retval.isEmpty() || isRefresh) {
+            return RestClient.service().listNew()
+                    .doOnNext(new putToSpAction(ctx, Constants.NEW_FILE_SP, Constants.NEW_VAL_SP));
+        } else {
+            return Observable.just(retval);
+        }
+    }
+
+    //TODO: add listAsk, listJob, and listShow
```

Comment:
But the `SharedPreferences` file name and val name key will be different (NEW_FILE_SP, TOP_FILE_SP, ASK_FILE_SP, etc).

Also, as the endpoint is different, the RestClient's method will be different too (listNewStories, listAskStories, listTopStories, etc).

Or do you mean to pass the file and val key name as this method's parameter and use switch-case for the RestClient's method?